Improve syntax docs (#33635)
hashicorp/terraform: website/docs/language/resources/syntax.mdx

as virtual networks, compute instances, or higher-level components such
as DNS records.

> To see how Terraform manages resources when applying a configuration, see

Author: reviewer note: removed extra paragraph and moved sentence up

> To see how Terraform manages resources when applying a configuration, see
> [Resource Behavior](/terraform/language/resources/behavior).

Reviewer: ```suggestion
For information about how Terraform manages resources after applying a configuration, refer to
[Resource Behavior](/terraform/language/resources/behavior).
```
Just a tweak for consistency with our style guide

## Resource Syntax

Resource declarations can include a number of advanced features, but only
a small subset are required for initial use. More advanced syntax features,
such as single resource declarations that produce multiple similar remote

> objects, are described later in this page.

Author: reviewer note: moved down

The resource type ("aws_instance") and name ("Web") together must be unique within a module because they

> serve as an identifier for a given resource.

Author: reviewer note: moved paragraph up to better explain the example

> A "resource" block declares a resource of a specific type 
> with a specific local name. The name is used to refer to this resource 
> in the same Terraform module but has no meaning outside that module's scope.
> 
> The resource type ("aws_instance") and name ("Web") together must be unique within a module because they
> serve as an identifier for a given resource.

Reviewer: ```suggestion
A `resource` block declares a resource of a specific type 
with a specific local name. Terraform uses the name when referring to the resource 
in the same module, but it has no meaning outside that module's scope.

In the following example,  the `aws_instance` resource type is named `web`. The resource type and name must be unique within a module because they serve as an identifier for a given resource.
```
A few more tweaks per our styles. In this case, use active voice and do not use quotation marks.

}
```

Within the block body (between `{` and `}`) are the configuration arguments
for the resource itself. The arguments often depend on the
resource type. In this example, both `ami` and `instance_type` are special 
arguments for [the `aws_instance` resource type](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance).

-> **Note:** Resource names must start with a letter or underscore, and may
contain only letters, digits, underscores, and dashes.

Resource declarations can include more advanced features, such as single 
resource declarations that produce multiple similar remote objects, but only

> a small subset is required for initial use. You will learn more later in this page.

Reviewer: ```suggestion
a small subset is required for initial use. Refer to [Resource Arguments](#resource-arguments) for additional information.
```
Is this the topic you are alluding to? Our writing guidelines instruct us to treat each section on a page as a contained module of information. Our research indicates that people rarely read a page from beginning to end, so treating the content like a narrative can be confusing. It's better to link directly to a related section.

Author: Got you. Not sure where this refers to tbh. Used the original texts. If you are sure that this the place to link to, lets go for this anchor. @trujillo-adam

Reviewer: Ah, I see. I think we should just remove the line, then. It was probably meant to refer to the article in its entirety, which is exactly counter the point I was making about writing it as a narrative. I'll modify my suggestion.

### Providers

A [provider](/terraform/language/providers/requirements) is a plugin for Terraform 
that offers a collection of resource types. Each resource type is implemented by a provider. A
provider provides resources to manage a single cloud or on-premises
infrastructure platform. Providers are distributed separately from Terraform, 

> but Terraform can automatically install most providers when initializing

Author: reviewer note: simplified texts

a working directory.

> To manage resources, a Terraform module must specify the required providers, see 
> [Provider Requirements](/terraform/language/providers/requirements).

Reviewer: ```suggestion
To manage resources, a Terraform module must specify the required providers. Refer to 
[Provider Requirements](/terraform/language/providers/requirements) for additional information.
```
Per our style guide, use "refer to" instead of "see".

Most providers need some configuration to access their remote API, 

> which is provided by the root module, see 
> [Provider Configuration](/terraform/language/providers/configuration)

Reviewer: ```suggestion
which is provided by the root module. Refer to 
[Provider Configuration](/terraform/language/providers/configuration) for additional information.
```

Based on a resource type's name, Terraform can usually determine which provider to use. 
By convention, resource type names start with their provider's preferred local name.

> When using multiple configurations of a provider (or non-preferred local provider names), 

Reviewer: ```suggestion
When using multiple configurations of a provider or non-preferred local provider names, 
```

you must use [the `provider` meta-argument](/terraform/language/meta-arguments/resource-provider) 

> to manually choose a provider configuration.

Author: reviewer note: removed list and directly link in the actual text

[expressions](/terraform/language/expressions) and other dynamic Terraform
language features.

> [Meta-Arguments](#meta-arguments) are defined by Terraform itself

Reviewer: ```suggestion
[Meta-arguments](#meta-arguments) are defined by Terraform
```

> and apply across all resource types. 

Author: reviewer note: moved link

### Documentation for Resource Types

Every Terraform provider has its own documentation, describing its resource
types and their arguments.

Some provider documentation is still part of Terraform's core documentation 
but the [Terraform Registry](https://registry.terraform.io/browse/providers) 

> is now the main home for all publicly available provider docs.

Author: reviewer note: just more simple now

> Some provider documentation is still part of Terraform's core documentation 
> but the [Terraform Registry](https://registry.terraform.io/browse/providers) 
> is now the main home for all publicly available provider docs.

Reviewer: ```suggestion
Some provider documentation is still part of Terraform's core documentation, 
but the [Terraform Registry](https://registry.terraform.io/browse/providers) 
is the main home for all publicly available provider docs.
```
Avoid using temporal language that can date the content.

[Terraform Registry](https://registry.terraform.io/browse/providers), which also
hosts their documentation. When viewing a provider's page on the Terraform
Registry, you can click the "Documentation" link in the header to browse its
documentation. Provider documentation on the registry is versioned, and you can
use the dropdown version menu in the header to switch which version's
documentation you are viewing.
Some provider documentation is still part of Terraform's core documentation 
but the [Terraform Registry](https://registry.terraform.io/browse/providers) 
is now the main home for all publicly available provider docs.

To browse the publicly available providers and their documentation, see
[the providers section of the Terraform Registry](https://registry.terraform.io/browse/providers).

-> **Note:** Provider documentation previously existed as part of Terraform's core documentation. Although some provider documentation
might still be hosted here, the Terraform Registry is now the main home for all
public provider docs.

## Resource Behavior

For more information about how Terraform manages resources when applying a
configuration, see

> [Resource Behavior](/terraform/language/resources/behavior).

Author: reviewer note: this has been move to the top

> When viewing a provider's page on the Terraform
> Registry, you can click the "Documentation" link in the header to browse its
> documentation, which is versioned. Use the dropdown version menu in the header 
> to switch the version.

Reviewer: ```suggestion
When viewing a provider's page on the Terraform
Registry, you can click the **Documentation** link in the header to browse its
documentation. The documentation is versioned. To choose a different version of the provider documentation, click on the version in the provider breadcrumbs to choose a version from the drop-down menu.
```
Some of these are existing style guide violations, so thanks for opening this PR and giving us the mandate to fix them.

}
```

> [Custom Condition Checks](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions) 

Reviewer: ```suggestion
[Custom condition checks](/terraform/language/expressions/custom-conditions#preconditions-and-postconditions) 
```

> can help capture assumptions, helping future maintainers 
> understand the configuration design and intent. They also return useful 

Reviewer: ```suggestion
can help capture assumptions so that future maintainers 
understand the configuration design and intent. They also return useful 
```

information about errors earlier and in context, helping consumers to diagnose 

> issues in their configuration.

Author: reviewer note: changed line length and moved links to the first mention of custom conditions
